src/checkmatelib/client.py

```diff
@@ -68,10 +75,30 @@ def check_url(  # pylint: disable=inconsistent-return-statements
         response.raise_for_status()
 
         if response.status_code == 204:
+            # No news is good news
             return None
 
         try:
             return BlockResponse(response.json())
 
         except ValueError as err:
             raise CheckmateServiceError("Unprocessable JSON response") from err
+
+    @classmethod
+    def _clean_url(cls, url):
+        """Clean the URL before we send it.
+
+        Checkmate will do this for itself, and doesn't trust our input, but
+        this allows us to fail fast for bad URLs, before Checkmate has a chance
+        to check them. We can also apply extra constraints.
```

Comment:
I think it would be worth noting here at a high level what it means for a URL to be "bad", since we're not using a standard definition. It looks to me like we're checking:

- Is this URL syntactically valid? IIRC `urlparse` is extremely liberal in what it will accept, are we just using its version of valid or something more constrained?
- Does the URL refer to a public host, or localhost?

What about non-HTTP URLs?

Reply:
We are more constrained than `urlparse`. To be honest I don't think it's really worth duplicating the details of our canonicalisation procedure (which is also in this repo) in comments here, but we could sign-post where to look.

We don't support non-HTTP URLs as we can't show them to you.

Reply:
Thanks for the clarification. What I was getting at here is that if there isn't a good existing definition of a "bad URL" that we can readily point to, it would be helpful for future us/developers to try to reify that a bit somehow (eg. some brief high-level comments or examples - no need to restate all the cases that the implementation checks for).